 WHATS_NEW | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/WHATS_NEW b/WHATS_NEW
index a1da4b7..82dd0c8 100644
--- a/WHATS_NEW
+++ b/WHATS_NEW
@@ -1,5 +1,11 @@
+Version 2.02.183 - 
+====================================
+  Use sync io if async io_setup fails, or use_aio=0 is set in config.
+
 Version 2.02.182 - 
 ==============================
+  Fix possible write race between last metadata block and the first extent.
+  Fix filtering of md 1.0 devices so they are not seen as duplicate PVs.
   Fix change of monitoring in clustered volumes.
   Fix lvconvert striped/raid0/raid0_meta -> raid6 regression.
   Add After=rbdmap.service to {lvm2-activation-net,blk-availability}.service.
