 conf/example.conf.in | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/conf/example.conf.in b/conf/example.conf.in
index 742812c..b37e0b2 100644
--- a/conf/example.conf.in
+++ b/conf/example.conf.in
@@ -903,6 +903,11 @@ global {
 	# This configuration option has an automatic default value.
 	# lvdisplay_shows_full_device_path = 0
 
+	# Configuration option global/use_aio.
+	# Use async I/O when reading and writing devices.
+	# This configuration option has an automatic default value.
+	# use_aio = 1
+
 	# Configuration option global/use_lvmetad.
 	# Use lvmetad to cache metadata and reduce disk scanning.
 	# When enabled (and running), lvmetad provides LVM commands with VG
