 WHATS_NEW           | 1 +
 lib/cache/lvmetad.c | 2 +-
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/WHATS_NEW b/WHATS_NEW
index 5de648c..9ddecbc 100644
--- a/WHATS_NEW
+++ b/WHATS_NEW
@@ -1,5 +1,6 @@
 Version 2.02.131 - 
 =====================================
+  Fix VG metadata comparison failure while scanning the VG for lvmetad.
   Ignore persistent cache if configuration changed. (2.02.127)
 
 Version 2.02.130 - 5th September 2015
diff --git a/lib/cache/lvmetad.c b/lib/cache/lvmetad.c
index b2e2f55..0fff65a 100644
--- a/lib/cache/lvmetad.c
+++ b/lib/cache/lvmetad.c
@@ -1174,7 +1174,7 @@ static struct volume_group *lvmetad_pvscan_vg(struct cmd_context *cmd, struct vo
 		if (!vgmeta_ret) {
 			vgmeta_ret = vgmeta;
 		} else {
-			if (!compare_config(vgmeta_ret->root, vgmeta->root)) {
+			if (compare_config(vgmeta_ret->root, vgmeta->root)) {
 				log_error("VG metadata comparison failed");
 				dm_config_destroy(vgmeta);
 				dm_config_destroy(vgmeta_ret);
