 lib/activate/activate.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lib/activate/activate.c b/lib/activate/activate.c
index 18f4b84..e38ab03 100644
--- a/lib/activate/activate.c
+++ b/lib/activate/activate.c
@@ -1864,6 +1864,8 @@ int monitor_dev_for_events(struct cmd_context *cmd, const struct logical_volume
 
 	if (!laopts)
 		laopts = &zlaopts;
+	else
+		mirr_laopts.read_only = laopts->read_only;
 
 	/* skip dmeventd code altogether */
 	if (dmeventd_monitor_mode() == DMEVENTD_MONITOR_IGNORE)
