commit dc1c96b8bebbd6bfe9b774fdb13f904f976f29bd
Author: David Teigland <teigland@redhat.com>
Date:   Thu Jul 2 13:45:38 2015 -0500

    lvmlockd: tech preview notice
---
 tools/toollib.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/tools/toollib.c b/tools/toollib.c
index 6b8ce22..b8edaea 100644
--- a/tools/toollib.c
+++ b/tools/toollib.c
@@ -981,6 +981,13 @@ int vgcreate_params_set_from_args(struct cmd_context *cmd,
 		vp_new->clustered = 0;
 
 	log_debug("Setting lock_type to %s", vp_new->lock_type);
+
+	if (is_lockd_type(vp_new->lock_type)) {
+		log_print("WARNING: shared lock type \"%s\" and lvmlockd are Technology Preview.", vp_new->lock_type);
+		log_print("For more information on Technology Preview features, visit:");
+		log_print("https://access.redhat.com/support/offerings/techpreview/");
+	}
+
 	return 1;
 }
 
